---
title: "Single Gene Report V1.2"
author: "Yilin Bai"
date: "2023-08-21"
output: 
  bookdown::html_document2:
    code_folding: show
    number_sections: no
    toc: yes
    toc_depth: 6
    toc_float: yes
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,message = FALSE, warning = FALSE)
```

## Reference

**Dataset:** Suvorov, Anton; Kim, Bernard; Wang, Jeremy; Armstrong, Ellie; Peede, David; D'Agostino, Emmanuel R. R.; et al. (2020). Widespread introgression across a phylogeny of 155 Drosophila genomes. figshare. Dataset. <https://doi.org/10.6084/m9.figshare.13264697.v1>

**Paper:** Suvorov A, Kim BY, Wang J, Armstrong EE, Peede D, D'Agostino ERR, Price DK, Waddell P, Lang M, Courtier-Orgogozo V, David JR, Petrov D, Matute DR, Schrider DR, Comeault AA. Widespread introgression across a phylogeny of 155 Drosophila genomes. Curr Biol. 2022 Jan 10;32(1):111-123.e5. doi: 10.1016/j.cub.2021.10.052. Epub 2021 Nov 16. PMID: 34788634; PMCID: PMC8752469.

## Target alignment

**Name:** EOG09150AOE

**Outgroup:** Anopheles

**Type:** Aligned DNA BUSCO loci

```{r}
library(Biostrings)


candidate <- 30
id <- as.character(candidate)
# "geneset_loc" is the folder where you save a set of multi-alignments
geneset_loc <- "../datasets/different_method/Squamates-Final_Alns/MAFFT-Auto-gt/"
gene_sets <- system2(command = "ls", args = geneset_loc, stdout = TRUE)
cat("Current gene file location:", geneset_loc, "\n")
cat("Number of gene:", length(gene_sets), "\n")
gene_file <- gene_sets[candidate]
gene_name <- gsub("(UCE-\\d+).*", "\\1", gene_file)
gene_path <- paste(geneset_loc, gene_file, sep = "")

# Set the path and filename for iqtree running result
prefix_single <- paste("./test/", "dmg_Single_", id, sep = "")
prefix_mix <- paste("./test/", "dmg_Mix_", id, sep = "")
# Set the command for both single and mix model
arg_single <- c("-s", gene_path, "-B", "1000", "--prefix", prefix_single)
arg_mix <- c("-s", gene_path, "-m", "ESTMIXNUM", "-mrate", "E,I,G,I+G,R,I+R", "-opt_qmix_criteria", "1", "--prefix", prefix_mix)

# The species tree file should be assigned manually
tree_path_species <- "../datasets/different_method/phylogenic_tree/Squamates/Rooted_MAFFT_Auto_GappyOut_Species_Tree_Scored.tre"
outgroup <- readLines("../datasets/different_method/Squamates-Final_Alns/Squamates-Taxon-Outgroup.txt")

cat("Current gene name:", gene_sets[candidate], "\n")
alignment <- readDNAMultipleAlignment(filepath = gene_path, format = "phylip")
```

## Command of model

```{r}
# Command of run one class model iq-tree
system2(command = "iqtree2", args = arg_single, stdout = FALSE)
# Command of run mixture class model iq-tree
system2(command = "iqtree2", args = arg_mix, stdout = FALSE)

```


## Alignment summary

```{r}
# plot 
# Visualization
# Extract a 100-bp window with lowest dominant base ratio
window_size <- 100

if (algn_length > window_size) {
  window_mean_dominant_base_ratio <- sapply(
    1:(algn_length - window_size + 1),
    function(i) {
      mean(maxmum_base_ratio[i:(i + window_size - 1)])
    }
  )
  start <- which.min(window_mean_dominant_base_ratio)
  end <- start + window_size - 1
  cat(paste("[Plot 1] Selected the window with the most sequence variation(", window_size, "bp) ï¼Œ", start, "-", end, "\n", sep = ""))
  alignment_fig <- ggmsa(
    start = start, end = end, alignment,
    char_width = 0.5, seq_name = F
  ) +
    geom_seqlogo()
} else {
  alignment_fig <- ggmsa(alignment, char_width = 0.5, seq_name = F) +
    geom_seqlogo()
}

alignment_fig
```

```{r}
# model
one_line <- readLines(paste(prefix_single, ".iqtree", sep = ""))
mix_line <- readLines(paste(prefix_mix, ".iqtree", sep = ""))

# length and Ntax
Length_line <- grep("Input data: ", one_line, value = TRUE)

Length_matches <- gregexpr("\\d+", Length_line)
Length_numbers <- regmatches(Length_line, Length_matches)[[1]]

# Convert the extracted strings to integers
Ntaxa <- as.integer(Length_numbers[1])
Length <- as.integer(Length_numbers[2])
```

```{r}
# percent_constant & percent_parsimony_informative

percent_constant_line <- grep("Number of constant sites: ", one_line, value = TRUE)
percent_constant <- sprintf("%.5s%%", regmatches(percent_constant_line, regexpr("\\d+\\.\\d+%\\s+", percent_constant_line)))
                            
percent_constant

parsimony_informative <- as.numeric(sub("Number of parsimony informative sites: ","\\1",grep("Number of parsimony informative sites: ", one_line, value = TRUE))) 
percent_parsimony_informative <- sprintf("%.2f%%", parsimony_informative / Length * 100)
percent_parsimony_informative

site_patterns <- as.numeric(sub("Number of distinct site patterns: ","\\1",grep("Number of distinct site patterns: ", one_line, value = TRUE)))
site_patterns
```

```{r}
# Create the data frame
alignment_data <- data.frame(
  "Locus_name" = "EOG09150AOE",
  "Length" = Length,
  "Ntaxa" = Ntaxa,
  "Percent_constant" = percent_constant,
  "Percent_parsimony_informative" = percent_parsimony_informative,
  "Site_patterns" = site_patterns
)
alignment_data
```


## Tree Topology

```{r}
library(ape)
library(phytools)
# Read the tree files
Mixture_tree <- read.tree(paste(prefix_single, ".treefile", sep = ""))
One_class_tree <- read.tree(paste(prefix_mix, ".treefile", sep = ""))
Species_tree <- read.tree("Data/mixture_class_single_gene/EOG09150A0E.fna.aln.tr.treefile")

```

### Phylograms plot
```{r fig.height=10, fig.width=16}
cladograms <- function(tree, y_scale = -5, subtitle = "") {
  ggtree(tree, cex = 0.8, layout = "slanted", branch.length = "none") +
    geom_tiplab(align = TRUE, size = 2) +
    geom_treescale(y = y_scale, color = "black", fontsize = 4) +
    labs(subtitle = subtitle) + 
    theme(legend.position = "bottom")
}

p1 <- cladograms(One_class_tree, subtitle = "Single Model")
p2 <- cladograms(Mixture_tree, subtitle = "Mix Model")

multiplot(p1, p2, ncol = 2)



phylograms <- function(tree, y_scale = -5, subtitle = "") {
  ggtree(tree, cex = 0.8, aes(color = branch.length - mean(branch.length))) +
    scale_color_continuous(
      high = "red", low = "black",
      name = "centralized(Branch Length)",
      guide = guide_colourbar(barwidth = 10)
    ) +
    geom_tiplab(align = TRUE, size = 2) +
    geom_treescale(y = y_scale, color = "black", fontsize = 4) +
    labs(subtitle = subtitle) +
    theme(legend.position = "bottom")
}

p1 <- phylograms(One_class_tree, subtitle = "Single Model")
p2 <- phylograms(Mixture_tree, subtitle = "Mix Model")

multiplot(p1, p2, ncol = 2)
```

### Cophylogeny plot

```{r fig.height=18, fig.width=10}
## create co-phylogenetic object
wasp.cophylo<-cophylo(Mixture_tree, One_class_tree)
## plot co-phylogenies
plot(wasp.cophylo,link.type="curved",link.lwd=4,
 link.lty="solid",link.col=make.transparent("red",
 0.25))
par(mar=c(5.1,4.1,4.1,2.1))
```

**Phylogenies inferred using these 3 approaches only differed in 2 trees:**

(1) D watanabei D punjabiensis was either have paraphyletic relationships to D. kikkawai and D. leontia or have paraphyletic relationships with D. seguy, D. nikananu, D. vulcana, D spaffchauvacae, D bocquet, D burlai, D. jambulina, D. bakoue

(2) D wassermani form monophyletic lineage sister to the D. acanthoptera or have paraphyletic relationships where D pachea is sister to the D. acanthoptera

(3) D paucipunta form monophyletic lineage sister to the D prolacticillia or have paraphyletic relationships with the D prolacticillia

### Table of topological distance metrics
```{r}
library(ape)
library(phangorn)

taxa_single <- One_class_tree$tip.label
taxa_mix <- Mixture_tree$tip.label
# Check if there are any different elements between taxa_single and taxa_mix
taxa_equal_model <- setequal(taxa_single, taxa_mix)
if (!taxa_equal_model){
  taxa_diff <- union(setdiff(taxa_single, taxa_mix),setdiff(taxa_mix, taxa_single))
  One_class_tree <- drop.tip(One_class_tree, taxa_diff)
  Mixture_tree <- drop.tip(Mixture_tree, taxa_diff)
  cat("The taxa of the two model trees are not equal, the taxa of the two trees are unified.\n")
}


taxa_species <- Species_tree$tip.label
taxa_subset <- intersect(taxa_single, taxa_species)
outgroup_subset <- intersect(taxa_subset, outgroup)
drop_taxa <- setdiff(taxa_species, taxa_subset)
tree_species_trimed <- drop.tip(tree_species, drop_taxa)


# Find the intersection of taxa
taxa_intersect <- intersect(One_class_tree$tip.label, Mixture_tree$tip.label)
taxa_subset <- intersect(taxa_intersect, tree_species$tip.label)
outgroup_subset <- intersect(taxa_subset, outgroup)

# Trim the species tree
tree_species_trimed <- drop.tip(tree_species, setdiff(tree_species$tip.label, taxa_subset))

# Root the trees if possible
rootable <- length(outgroup_subset) > 0
if (rootable) {
  tryCatch({
    One_class_tree <- root(One_class_tree, outgroup_subset)
    Mixture_tree <- root(Mixture_tree, outgroup_subset)
    tree_species_trimed <- root(tree_species_trimed, outgroup_subset)
  }, warning = function(w) {}, 
  error = function(e) {
    print("specified outgroup is not monophyletic")
  })

}

# Check if all trees are rooted
all_rooted <- rootable && is.rooted(One_class_tree) && is.rooted(Mixture_tree) && is.rooted(tree_species)


# Placeholder data (replace these with your actual data)
tree_names <- c("One_class", "Mixture", "Species")
tree_files <- list(
  One_class = One_class_tree,
  Mixture = Mixture_tree,
  Species = tree_species_trimed
)

# Create an empty data frame to store the results
result_df <- data.frame(
  metric = character(0),
  tree_1 = character(0),
  tree_2 = character(0),
  gene_name = character(0),
  RF_distance = numeric(0),
  nRF_distance = numeric(0),
  wRF_distance = numeric(0),
  KF_distance = numeric(0),
  PD_distance = numeric(0),
  wPD_distance = numeric(0),
  geodesic_distance = numeric(0)
)

# Loop through the combinations of gene names and tree pairs

tree_combinations <- combn(length(tree_files), 2, simplify = FALSE)
for (comb in tree_combinations) {
    i <- comb[1]
    j <- comb[2]
    
    tree1 <- tree_files[[i]]
    tree2 <- tree_files[[j]]
    tree1_name <- tree_names[i]
    tree2_name <- tree_names[j]
    
    # Calculate distances
    RF_dist <- RF.dist(tree1, tree2, normalize = FALSE, check.labels = TRUE, rooted = all_rooted)
    nRF_dist <- RF.dist(tree1, tree2, normalize = TRUE, check.labels = TRUE, rooted = all_rooted)
    wRF_dist <- wRF.dist(tree1, tree2, normalize = FALSE, check.labels = TRUE, rooted = all_rooted)
    KF_dist <- KF.dist(tree1, tree2, check.labels = TRUE, rooted = all_rooted)
    PD_dist <- path.dist(tree1, tree2, check.labels = TRUE, use.weight = FALSE)
    wPD_dist <- path.dist(tree1, tree2, check.labels = TRUE, use.weight = TRUE)
    
    # If the trees are rooted calculate geodesic distance
    if (all_rooted) {
      geodesic_dist <- dist.multiPhylo(c(tree1,tree2), method = "geodesic", outgroup = outgroup_subset)
    } else {
      geodesic_dist <- NA
    }
  
    # Add the results to the data frame
    result_df <- rbind(result_df, data.frame(
    gene_name = gene_name,
    tree_1 = tree1_name,
    tree_2 = tree2_name,
    RF_distance = RF_dist,
    nRF_distance = nRF_dist,
    wRF_distance = wRF_dist,
    KF_distance = KF_dist,
    PD_distance = PD_dist,
    wPD_distance = wPD_dist,
    geodesic_distance = geodesic_dist
      ))

}

# Print the resulting data frame
knitr::kable(result_df, digits = 4, caption = "Tree distance Summary")

```

## Branch Lengths

### Summary parameters table

```{r}
# tree length & Sum_int & prop_int
one_length <- sub("Total tree length \\(sum of branch lengths\\): ","\\1",grep("Total tree length \\(sum of branch lengths\\): ", one_line, value = TRUE))
mix_length <- sub("Total tree length \\(sum of branch lengths\\): ","\\1",grep("Total tree length \\(sum of branch lengths\\): ", mix_line, value = TRUE))

one_sum_int <- sub("Sum of internal branch lengths: ([0-9.]+).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", one_line, value = TRUE))
mix_sum_int <- sub("Sum of internal branch lengths: ([0-9.]+).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", mix_line, value = TRUE))

one_prop_int <- sub(" of tree length", "", sub(".*\\(([^)]+)\\).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", one_line, value = TRUE)))
mix_prop_int <- sub(" of tree length", "", sub(".*\\(([^)]+)\\).*","\\1",grep("Sum of internal branch lengths: ([0-9.]+).*", mix_line, value = TRUE)))
```

```{r}
# Provided Newick tree string
Mixture_txt <- readLines("Data/mixture_class_single_gene/mix_class_EOG09150A0E.treefile")
One_class_txt <- readLines("Data/one_class_single_gene/one_class_EOG09150A0E.treefile")

branch_len_mix <- as.numeric(gsub("([0-9.]+).*", "\\1", strsplit(Mixture_txt, ":")[[1]])[-1])
branch_len_one <- as.numeric(gsub("([0-9.]+).*", "\\1", strsplit(One_class_txt, ":")[[1]])[-1])

sum_branch_len_one <- summary(branch_len_one)
sum_branch_len_mix <- summary(branch_len_mix)
```

```{r}
# Create a dataframe with the provided summary statistics
df <- data.frame(
  Tree_Length = c(one_length, mix_length),
  Sum_int = c(one_sum_int, mix_sum_int),
  prop_int = c(one_prop_int, mix_prop_int),
  min = c(sum_branch_len_one[1], sum_branch_len_mix[1]),
  Qu_1st = c(sum_branch_len_one[2], sum_branch_len_mix[2]),
  Median = c(sum_branch_len_one[3], sum_branch_len_mix[3]),
  Mean = c(sum_branch_len_one[4], sum_branch_len_mix[4]),
  Qu_3rd = c(sum_branch_len_one[5], sum_branch_len_mix[5]),
  Max = c(sum_branch_len_one[6], sum_branch_len_mix[6]),
  gene_name = rep(gene_names,2),
  row.names = c("One model", "Mixture model"))
# Print the dataframe
print(df)
```

### Faceted histogram

```{r}
# Assuming you have loaded the 'ggplot2' package and have the necessary data

data <- data.frame(
  model = rep(c("One class", "Mixture"), each = length(branch_len_one)),
  branch_length = c(branch_len_one, branch_len_mix)
)

# Create a faceted histogram
library("ggplot2")
ggplot(data, aes(x = branch_length)) +
  geom_histogram(binwidth = 0.1, fill = "blue", color = "black") +
  facet_grid(model ~ ., scales = "free_y") +  # Facet by 'model', free y-axis scales
  labs(x = "Branch Length", y = "Frequency") +
  theme_minimal()+
  scale_x_log10()
```

### ECDF plot

```{r}
# Create an ECDF plot
ggplot(data, aes(x = branch_length, color = model)) +
  stat_ecdf(geom = "step") +
  labs(x = "Branch Length", y = "ECDF") +
  scale_color_manual(values = c("One class" = "blue", "Mixture" = "red")) +
  theme_minimal() +
  theme(legend.position = "top")+
  scale_x_log10()
```

## Models

### Description
```{r}
source("./iqtree_info.r")
sum_one <- summarise_iqtree(one_line)
sum_mix <- summarise_iqtree(mix_line)
cat("The phylogenic acnalysis result for one model:\n")
short_description(sum_one)
cat("---------------------------------------------
The phylogenic analysis result for mix model:\n")
short_description(sum_mix)

model_summary_table <- rbind(summary_table(sum_one),summary_table(sum_mix))
model_summary_table <- cbind(Model = c("One_class","Mixture"), model_summary_table)
best_model <- model_summary_table$Model[[which.min(model_summary_table$BIC)]]
model_summary_table <- cbind(model_summary_table, Best = rep(best_model, 2))
knitr::kable(model_summary_table, digits = 4, caption = "Model Summary")
```


